### PR TITLE
cnf-tests: SRIOV/NUMA consume all VFs

### PR DESCRIPTION
**Pull request by zeeke**
Add a test case where all Virtual Function that belongs to an SriovNetworkNodePolicy with `ExcludeTopology = true` are consumed. Test logic verifies further pods deployments get "Insufficient resource" errors, then frees some VFs and pod can be scheduled again.

cc @gregkopels

### CODE REVIEW

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go [hunk 1]
```diff
@@ -142,6 +142,58 @@ var _ = Describe("[sriov] NUMA node alignment", Ordered, func() {
...
 	})
+
+	It("Utilize all available VFs then create a pod with guaranteed CPU and excludeTopology set to True", func() {
+		podDefinition := pods.DefinePod(sriovnamespaces.Test)
+		podDefinition = pods.RedefineWithGuaranteedQoS(podDefinition, "2", "500Mi")
+		podDefinition = pods.RedefinePodWithNetwork(podDefinition, "test-numa-0-exclude-topology-true-network")
+
+		By("Utilize all available VFs")
+		var expectedRunningPods []*corev1.Pod
+		for i := 0; i < 4; i++ {
```
Comment (SchSeba) [inline cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:153]: can you just create a pod requesting 5 devices?

Comment (SchSeba) [reply cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:153]: also please use the numVF variable :)

Comment (zeeke) [reply cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:153]: I have to implement a step where all VFs are in use, so at least I can do the following:

```
create a pod with 4 VFs
assert it's running

create a pod with 1 VF (or more)
assert it raises an error
```

If I create a variable `numVFs` , I still need a loop to create the network annotation:

```go
                ...
                podDefinition := pods.DefinePod(sriovnamespaces.Test)
		podDefinition = pods.RedefineWithGuaranteedQoS(podDefinition, "2", "500Mi")

		numVFs := 4
		By("Utilize all available VFs")
		expectedRunningPod := podDefinition.DeepCopy()
		useAllVFsNetworkSpec := []string{}
		for i := 0; i < numVFs; i++ {
			useAllVFsNetworkSpec = append(useAllVFsNetworkSpec, "test-numa-0-exclude-topology-true-network")
		}
		expectedRunningPod = pods.RedefinePodWithNetwork(expectedRunningPod, strings.Join(","))

		expectedRunningPod, err := client.Client.Pods(sriovnamespaces.Test).
			Create(context.Background(), podDefinition, metav1.CreateOptions{})
		Expect(err).ToNot(HaveOccurred())

		Eventually(func(g Gomega) {
			actualPod, err := client.Client.Pods(sriovnamespaces.Test).Get(context.Background(), expectedRunningPod.Name, metav1.GetOptions{})
			g.Expect(err).ToNot(HaveOccurred())
			g.Expect(actualPod.Status.Phase).To(Equal(corev1.PodRunning))
			g.Expect(actualPod.Status.QOSClass).To(Equal(corev1.PodQOSGuaranteed))
		}, 30*time.Second, 1*time.Second).Should(Succeed())

		By("Create a failing pod that uses a VF")
		failingPod := podDefinition.DeepCopy()
		failingPod = pods.RedefinePodWithNetwork(failingPod, "test-numa-0-exclude-topology-true-network")
		failingPod, err = client.Client.Pods(sriovnamespaces.Test).
			Create(context.Background(), failingPod, metav1.CreateOptions{})
		Expect(err).ToNot(HaveOccurred())
                ...
```

Is that what you had in mind?

Comment (SchSeba) [reply cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:153]: yep I think is nicer then having 5 pods.
we have the same in another sriov test https://github.com/openshift/sriov-network-operator/blob/master/test/conformance/tests/test_sriov_operator.go#L837

**Comment by liornoy** [comment]
lgtm

**Comment by zeeke** [comment]
Added a comment to better explain why we gather cpuset from multiple file paths

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go [hunk 2]
```diff
@@ -280,6 +281,61 @@ var _ = Describe("[sriov] NUMA node alignment", Ordered, func() {
 			g.Expect(actualPod.Status.Message).To(ContainSubstring("Resources cannot be allocated with Topology locality"))
 		}, 30*time.Second, 1*time.Second).Should(Succeed())
 	})
+
+	It("Utilize all available VFs then create a pod with guaranteed CPU and excludeTopology set to True", func() {
+		podDefinition := pods.DefinePod(sriovnamespaces.Test)
+		podDefinition = pods.RedefineWithGuaranteedQoS(podDefinition, "2", "500Mi")
```
Comment (fedepaol) [inline cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:287]: let's avoid shadowing names. 
I'd name:
barePod
withQos
withNetwork...
and so on. Using the same names is a bit confusing / prone to errors

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go [hunk 3]
```diff
@@ -280,6 +281,61 @@ var _ = Describe("[sriov] NUMA node alignment", Ordered, func() {
...
+
+		Eventually(pods.GetStringEventsForPodFn(client.Client, failingPod), 30*time.Second, 1*time.Second).
+			Should(ContainSubstring("Insufficient openshift.io/testNuma0NIC1ExcludeTopoplogyTrue"))
+
+		By("Release all VFs by deleting the running pod")
+		err = client.Client.Pods(sriovnamespaces.Test).
+			Delete(context.Background(), expectedRunningPod.Name, metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Create a pod to verify VFs are usable")
```
Comment (fedepaol) [inline cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:324]: should the failing pod be running too here?

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go [hunk 4]
```diff
@@ -280,6 +281,61 @@ var _ = Describe("[sriov] NUMA node alignment", Ordered, func() {
...
 		}, 30*time.Second, 1*time.Second).Should(Succeed())
 	})
+
+	It("Utilize all available VFs then create a pod with guaranteed CPU and excludeTopology set to True", func() {
+		podDefinition := pods.DefinePod(sriovnamespaces.Test)
+		podDefinition = pods.RedefineWithGuaranteedQoS(podDefinition, "2", "500Mi")
+
+		numVFs := 4
+		By("Utilize all available VFs")
+		expectedRunningPod := podDefinition.DeepCopy()
```
Comment (fedepaol) [inline cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:291]: name this by property, not by the expectation, like "podWithAllVfs" and document the expectations in `By`

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go (hunk 2 above)
Comment (zeeke) [reply cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:287]: Good point

File: cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go (hunk 3 above)
Comment (zeeke) [reply cnf-tests/testsuites/e2esuite/dpdk/numa_node_sriov.go:324]: good point. I tested locally and it actually gets started. I can simplify the test by removing a pod from the game

**Comment by gkopels** [comment]
lgtm

**Comment by zeeke** [comment]
`ci/prow/e2e-gcp-ovn` failure is not related to this PR:

```
[rfe_id:27368][performance] Network latency parameters adjusted by the Node Tuning Operator [test_id:28467][crit:high][vendor:cnf-qe@redhat.com][level:acceptance] Should contain configuration injected through the openshift-node-performance profile
```

/override ci/prow/e2e-gcp-ovn

**Pull request merged**